### PR TITLE
fix(custom-resources): cross-environment call fails in opt-in region

### LINKED ISSUE #26562: [AwsCustomResource]: (assumeRoleArn defined in non-opt-in region while assume in opt-in region cause permission issue)
### Describe the bug

for AwsCustomResource, the AwsSdkCall have assumeRoleArn that we can assume to proceed the SdkCall. 

But the default sts endpoint is set to regional, and default region in our case an opt-in region. However, the role is created/defined under a root account in non-opt-in region(we cannot enable all opt-in region for that account). With the incorrect sts endpint point the assumeRole failed with permission issue.

### Expected Behavior

sts Assume Role success and AwsSdkCall can proceed with success response.

### Current Behavior
```
Error [CredentialsError]: Missing credentials in config, if using AWS_CONFIG_FILE, set AWS_SDK_LOAD_CONFIG=1
    at Request.extractError (/var/runtime/node_modules/aws-sdk/lib/protocol/query.js:50:29)
    at Request.callListeners (/var/runtime/node_modules/aws-sdk/lib/sequential_executor.js:106:20)
    at Request.emit (/var/runtime/node_modules/aws-sdk/lib/sequential_executor.js:78:10)
    at Request.emit (/var/runtime/node_modules/aws-sdk/lib/request.js:686:14)
    at Request.transition (/var/runtime/node_modules/aws-sdk/lib/request.js:22:10)
    at AcceptorStateMachine.runTo (/var/runtime/node_modules/aws-sdk/lib/state_machine.js:14:12)
    at /var/runtime/node_modules/aws-sdk/lib/state_machine.js:26:10
    at Request.<anonymous> (/var/runtime/node_modules/aws-sdk/lib/request.js:38:9)
    at Request.<anonymous> (/var/runtime/node_modules/aws-sdk/lib/request.js:688:12)
    at Request.callListeners (/var/runtime/node_modules/aws-sdk/lib/sequential_executor.js:116:18) {
  code: 'CredentialsError',
  time: 2023-07-19T01:26:20.624Z,
  requestId: '***********',
  statusCode: 403,
  retryable: false,
  retryDelay: 30.07614769919884,
  originalError: {
    message: 'Could not load credentials from ChainableTemporaryCredentials',
    code: 'CredentialsError',
    time: 2023-07-19T01:26:20.624Z,
    requestId: '*****************',
    statusCode: 403,
    retryable: false,
    retryDelay: 30.07614769919884,
    originalError: {
      message: 'User: arn:aws:sts::*********:assumed-role/********* is not authorized to perform: sts:AssumeRole on resource: arn:aws:iam::**********:role/*******',
      code: 'AccessDenied',
      time: 2023-07-19T01:26:20.565Z,
      requestId: '******',
      statusCode: 403,
      retryable: false,
      retryDelay: 30.07614769919884
    }
  }
}
```
### Reproduction Steps

Create an AwsCustomResource in opt-in region to assume a Role define in an account which did not enable this opt-in region.

### Possible Solution

We tried all combinations of region vs stsEndpoint:

case 1, default region(opt-in region) with global sts endpoint. -> FAILED
case 2, non opt-in region with global sts endpoint. ->  SUCCEED
case 3, default region(opt-in region) with regional sts endpoint. ->  FAILED
case 4, non opt-in region with regional sts endpoint. ->  SUCCEED

So in either cases, we need to override the default region to a non-opt-in region!! So requesting to expose this sts region option to the user. the AwsSdkCall do have a region option, but the region is NOT used for sts assume role.

### Additional Information/Context

_No response_

### CDK CLI Version

2.73.0

### Framework Version

_No response_

### Node.js Version

18

### OS

AL2

### Language

Typescript

### Language Version

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by scanlonp**
Currently, the region parameter in `AwsCustomResource` only controls where the action is performed. If a role needs to be assumed, the `assumeRole` call is made from the region the stack is deployed into. This presents a problem if the stack is deployed into an opt-in region, and the role being assumed lives in a separate stack in an account without the opt-in region enabled. 

This change makes the `assumeRole` call and the sdk call performed in the same region. Therefore, to solve the above problem, pass any region that is enabled for the account that owns the role to be assumed.

Closes #26562.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Review by rix0rrr — APPROVED** [review]
Conditionally approved!

File: packages/@aws-cdk-testing/framework-integ/test/custom-resources/test/aws-custom-resource/integ.cross-account-assumeRole.ts [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+ *      and needs to set trust permissions for account B (987654321)
+ *      - `cdk bootstrap --trust 987654321 --cloudformation-execution-policies 'arn:aws:iam::aws:policy/AdministratorAccess'`
+ *      - assuming this is the default profile for aws credentials
+ *   b. Account B (987654321) should be bootstrapped for us-east-1 and af-south-1
+ *     - note Account B needs to have af-south-1 enabled as it is an opt-in region
+ *     - assuming this account is configured with the profile 'cross-account' for aws credentials
+ *
+ * 2. Set environment variables
+ *   a. `export CDK_INTEG_ACCOUNT=123456789`
+ *   b. `export CDK_INTEG_CROSS_ACCOUNT=987654321`
```
Comment (kaizencc) [inline packages/@aws-cdk-testing/framework-integ/test/custom-resources/test/aws-custom-resource/integ.cross-account-assumeRole.ts:21]:
```suggestion
 *   a. `export CDK_INTEG_ACCOUNT=123456789012`
 *   b. `export CDK_INTEG_CROSS_ACCOUNT=234567890123`
```

https://github.com/aws/aws-cdk/blob/main/.gitallowed

File: packages/@aws-cdk-testing/framework-integ/test/custom-resources/test/aws-custom-resource/integ.cross-account-assumeRole.ts [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+import { IntegTest } from '@aws-cdk/integ-tests-alpha';
+import { AwsCustomResource, AwsCustomResourcePolicy, PhysicalResourceId } from 'aws-cdk-lib/custom-resources';
+
+/**
+ * Notes on how to run this integ test
+ * (All regions are flexible, my testing used account A with af-south-1 not enabled)
+ *
+ * 1. Configure Accounts
+ *   a. Account A (123456789) should be bootstrapped for us-east-1
+ *      and needs to set trust permissions for account B (987654321)
```
Comment (kaizencc) [inline packages/@aws-cdk-testing/framework-integ/test/custom-resources/test/aws-custom-resource/integ.cross-account-assumeRole.ts:12]:
```suggestion
 *   a. Account A (123456789012) should be bootstrapped for us-east-1
 *      and needs to set trust permissions for account B (234567890123)
```

File: packages/@aws-cdk-testing/framework-integ/test/custom-resources/test/aws-custom-resource/integ.cross-account-assumeRole.ts [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+ * 3. Run the integ test (from the @aws-cdk-testing/framework-integ/test directory)
+ *   a. Get temporary console access credentials for account B
+ *     - `yarn integ custom-resources/test/aws-custom-resource/integ.cross-account-assumeRole.js`
+ *   b. Fall back if temp credentials do not work (account info may be in snapshot)
+ *     - `yarn integ custom-resources/test/aws-custom-resource/integ.cross-account-assumeRole.js --profiles cross-account`
+*/
+
+const app = new cdk.App();
+
+const account = process.env.CDK_INTEG_ACCOUNT || process.env.CDK_DEFAULT_ACCOUNT;
```
Comment (kaizencc) [inline packages/@aws-cdk-testing/framework-integ/test/custom-resources/test/aws-custom-resource/integ.cross-account-assumeRole.ts:32]:
```suggestion
const account = process.env.CDK_INTEG_ACCOUNT || process.env.CDK_DEFAULT_ACCOUNT || '123456789012';
```

File: packages/@aws-cdk-testing/framework-integ/test/custom-resources/test/aws-custom-resource/integ.cross-account-assumeRole.ts [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+// As the integ-runner doesnt provide a default cross account, we make our own.
+const crossAccount = process.env.CDK_INTEG_CROSS_ACCOUNT || '987654321';
```
Comment (kaizencc) [inline packages/@aws-cdk-testing/framework-integ/test/custom-resources/test/aws-custom-resource/integ.cross-account-assumeRole.ts:35]:
```suggestion
const crossAccount = process.env.CDK_INTEG_CROSS_ACCOUNT || '234567890123';
```

File: packages/aws-cdk-lib/custom-resources/README.md [hunk 5]
```diff
@@ -667,3 +667,28 @@ new cr.AwsCustomResource(this, 'GetParameter', {
   },
 });
 ```
+
+#### Making Cross Account Calls
+
+Example of making a cross account call using an assumed role. If deploying the custom resource in a region where the cross account role is not defined (i.e. an opt-in region that is not enabled in the account owning the role), set the region parameter to a region enabled in that account.
+
+```ts
+const crossAccountRoleArn = '...' // arn of role deployed in separate account
```
Comment (kaizencc) [inline packages/aws-cdk-lib/custom-resources/README.md:676]:
```suggestion
const crossAccountRoleArn = 'arn:aws:iam:account:000000000000' // arn of role deployed in separate account
```

somethign like this

**Comment by scanlonp** [comment]
Moving my personal checklist to comments.

TODO:
- [x] Specify accounts for the integ test to provide coverage for cross account and opt-in / non-opt-in region relationship

- reference [cross account integ test](https://github.com/aws/aws-cdk/blob/4fd510ec75f03e3e674911cd347fa96ea3edd93a/packages/%40aws-cdk-testing/framework-integ/test/aws-events/test/integ.cross-account-rule.ts)

- [x] ~Add unit tests to verify the configuration of the sts `assumeRole` call~ covered by integ test

Future work, not needed to merge
- [ ] Roll this change into addressing #26593

**Pull request merged**